ptvsd/wrapper.py

```diff
@@ -508,6 +508,68 @@ def get_sort_key(o):
         return self.variables + self.single_underscore + self.double_underscore + self.dunder  # noqa
 
 
+class ModulesManager(object):
+    def __init__(self, proc):
+        self.module_id_to_details = {}
+        self.path_to_module_id = {}
+        self._lock = threading.Lock()
+        self.proc = proc
+        self._next_id = 1
+
+    def add_or_get_from_path(self, module_path):
+        try:
+            module_id = self.path_to_module_id[module_path]
+            return self.module_id_to_details[module_id]
+        except KeyError:
+            pass
+        return self.__add_and_get_module_from_path(module_path)
+
+    def __get_platform_file_path(self, path):
+        if platform.system() == 'Windows':
+            return path.lower()
+        return path
+
+    def __add_and_get_module_from_path(self, module_path):
+        mpath = self.__get_platform_file_path(module_path);
+        for _, value in list(sys.modules.items()):
+            try:
+                path = self.__get_platform_file_path(value.__file__)
+            except AttributeError:
+                path = None
+
+            if path and mpath == path:
+                with self._lock:
+                    module_id = self._next_id
+                    self._next_id += 1
+
+                    self.path_to_module_id[module_path] = module_id
+
+                    module = {
+                        'id': module_id,
+                        'name': value.__name__,
```

Comment:
`__qualname__` is also interesting when it's available (and can then be used in `__format_frame_name`).